### PR TITLE
FIT digitization and reconstruction  

### PR DESCRIPTION
**Pull request by AllaMaevskaya**

### CODE REVIEW

File: Detectors/FIT/simulation/src/Detector.cxx [hunk 1]
```diff
@@ -313,8 +304,6 @@ Bool_t Detector::ProcessHits(FairVolume* v)
 HitType* Detector::AddHit(float x, float y, float z, float time, float energy, Int_t trackId, Int_t detId)
 {
   mHits->emplace_back(x, y, z, time, energy, trackId, detId);
-  auto stack = (o2::Data::Stack*)fMC->GetStack();
-  stack->addHit(GetDetId());
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/Detector.cxx]: These 2 lines were added 2 days ago as an important fix and they must stay. I don't understand the refactoring of not using the AddHit function anymore. This is a nice encapsulation but for performance reasons we should probably inline it.

File: Detectors/FIT/simulation/src/Detector.cxx [hunk 2]
```diff
@@ -302,7 +291,9 @@ Bool_t Detector::ProcessHits(FairVolume* v)
     if (iPart == 50000050) // If particles is photon then ...
     {
       if (RegisterPhotoE(etot)) {
-        AddHit(x, y, z, time, enDep, trackID, detID);
+	HitType newhit(x, y, z, time, enDep, trackID, detID);
+	mHits->push_back(newhit);
+	  //AddHit(x, y, z, time, enDep, trackID, detID);
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/Detector.cxx]: I suggest to keep it the way it was.

File: Detectors/FIT/simulation/include/FITSimulation/Detector.h [hunk 3]
```diff
@@ -59,10 +59,6 @@ class Detector : public o2::Base::DetImpl<Detector>
 
   /// Default constructor
   Detector() = default;
-
-  /// Clone this object (used in MT mode only)
-  FairModule* CloneModule() const override;
```
Comment (sawenzel) [inline Detectors/FIT/simulation/include/FITSimulation/Detector.h]: why is this taken away? This is necessary for G4-MT mode.

File: Detectors/FIT/simulation/src/Detector.cxx [hunk 4]
```diff
@@ -26,7 +26,6 @@
 #include <sstream>
 #include "FITBase/Geometry.h"
 #include "FITSimulation/Detector.h"
-#include "SimulationDataFormat/Stack.h"
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/Detector.cxx]: must stay

File: Detectors/FIT/reconstruction/src/CollisionTimeRecoTask.cxx [hunk 5]
```diff
@@ -18,9 +18,9 @@
 #include "FairLogger.h"      // for LOG
 #include "FairRootManager.h" // for FairRootManager
 
-ClassImp(o2::fit::CollisionTimeRecoTask);
+ClassImp(o2::fit::CollisionTimeRecoTask)
```
Comment (sawenzel) [inline Detectors/FIT/reconstruction/src/CollisionTimeRecoTask.cxx]: ';' must stay otherwise formatting of the next line breaks

File: Detectors/FIT/reconstruction/include/FITReconstruction/RecPoints.h [hunk 6]
```diff
@@ -11,36 +11,44 @@ namespace o2
...
     : mCollisionTime(collisiontime),
       mVertex(vertex),
       mTimeAmp(std::move(timeamp))
   {
   }
   ~RecPoints() = default;
 
-  void FillFromDigits(const std::vector<Digit>& digits);
+  //void FillFromDigits(const std::vector<Digit>& digits);
+  void FillFromDigits(const Digit* digit);
```
Comment (sawenzel) [inline Detectors/FIT/reconstruction/include/FITReconstruction/RecPoints.h]: do you need a pointer here? it is generally preferred to pass references.

File: Detectors/FIT/reconstruction/include/FITReconstruction/CollisionTimeRecoTask.h [hunk 7]
```diff
@@ -24,16 +24,18 @@ class CollisionTimeRecoTask : public FairTask
 {
  public:
   CollisionTimeRecoTask();
-  ~CollisionTimeRecoTask() override;
+  virtual ~CollisionTimeRecoTask();
```
Comment (sawenzel) [inline Detectors/FIT/reconstruction/include/FITReconstruction/CollisionTimeRecoTask.h]: the correct way is to use `override` as this is overriding a virtual function defined in the base class.

File: Detectors/FIT/simulation/src/Digitizer.cxx [hunk 8]
```diff
@@ -10,54 +10,234 @@
...
+  digit->setTime(digit_timeframe);
+  digit->setBC(mEventID);
+  digit->setTriggers(Is_A, Is_C, Is_Central, Is_SemiCentral, Is_Vertex);
+
+  std::vector<ChannelData> mChDgDataArr;
+  //  LOG(DEBUG) << "nMCP : :IsA : hit nPe : hit mTime : sig MIP : sig mTime" << FairLogger::endl;
+  for (Int_t ch_iter = 0; ch_iter < nMCPs; ch_iter++) {
+    if (ch_signal_MIP[ch_iter] > CFD_trsh_mip) {
+      Float_t smeared_time = gRandom->Gaus(ch_signal_time[ch_iter], 0.050) + BCEventTime;
+      mChDgDataArr.emplace_back(ChannelData{ ch_iter, smeared_time, ch_signal_MIP[ch_iter]} );
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/Digitizer.cxx]: the CI is reporting a compilation error here due to a narrowing conversion of double to float in ch_signal_MIP.
So either a cast needs to be applied or store ch_signal_MIP in float as well.

Moreover, if ChannelData gets a constructor this line can be simplified to

```
mChDgDataArr.emplace_back( ch_iter, smeared_time, ch_signal_MIP[ch_iter] );
```

File: Detectors/FIT/simulation/src/Digitizer.cxx [hunk 9]
```diff
@@ -10,54 +10,234 @@
...
+  // digit->printStream( std::cout );
+  
+  
+  LOG(DEBUG) << "======================================\n\n" << FairLogger::endl;
+  // --------------------------------------------------------------------------
+
+
+
+
+//  Int_t amp[nMCPs] = {};
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/Digitizer.cxx]: do you need to keep this commented code?

File: Detectors/FIT/simulation/src/DigitizerTask.cxx [hunk 10]
```diff
@@ -24,10 +24,13 @@ using namespace o2::fit;
...
 DigitizerTask::~DigitizerTask()
 {
-  if (mDigitsArray) {
-    mDigitsArray->clear();
-    delete mDigitsArray;
-  }
+//  if (mDigitsArray) {
+//    mDigitsArray->clear();
+//    delete mDigitsArray;
+//  }
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/DigitizerTask.cxx]: what is the reason for this change?

File: Detectors/FIT/simulation/src/Detector.cxx [hunk 11]
```diff
@@ -292,7 +281,7 @@ Bool_t Detector::ProcessHits(FairVolume* v)
     fMC->TrackPosition(x, y, z);
     fMC->CurrentVolID(quadrant);
     fMC->CurrentVolOffID(1, mcp);
-    float time = fMC->TrackTime() * 1.0e12;
+    float time = fMC->TrackTime() * 1.0e9;
```
Comment (sawenzel) [inline Detectors/FIT/simulation/src/Detector.cxx]: I think it would be good explaining the factor 1E-09 in a comment: `// We are converting from xxx to yyy`

**Comment by AllaMaevskaya** [comment]
Dear Sandro,
thank you for precise  look ! I will try to fix all these bugs.

Best regards
Alla

**Comment by sawenzel** [comment]
I took the liberty to provide an alternative PR #1334 which solves some issues (failing tests / wrongly added files / clean and squashed commits ). Please take a look. If approved, this can be merged quickly so that development can proceed.
